Index: lcl/lclproc.pas
===================================================================
--- lcl/lclproc.pas	(revision 52799)
+++ lcl/lclproc.pas	(working copy)
@@ -27,7 +27,7 @@
 uses
   {$IFDEF Darwin}MacOSAll, {$ENDIF}
   {$IFDEF win32}
-  Win9xWsManager, // Support for Lower/UpperWideStringProc on Win9x, also used by some Utf8 string handling functions
+  //Win9xWsManager, // Support for Lower/UpperWideStringProc on Win9x, also used by some Utf8 string handling functions
   {$ENDIF}
   {$IFnDEF WithOldDebugln} LazLogger, {$ENDIF}
   Classes, SysUtils, Math, TypInfo, Types, FPCAdds, AvgLvlTree, LazFileUtils,
