Index: lcl/interfaces/cocoa/cocoascrollers.pas
===================================================================
--- lcl/interfaces/cocoa/cocoascrollers.pas	(revision 63435)
+++ lcl/interfaces/cocoa/cocoascrollers.pas	(working copy)
@@ -50,9 +50,7 @@
     procedure resetCursorRects; override;
     function lclClientFrame: TRect; override;
     function lclContentView: NSView; override;
-    procedure setDocumentView(aView:
-      {$if FPC_FULLVERSION < 30200}NSView{$else}id{$endif}
-    ); override;
+    procedure setDocumentView(aView:NSView); override;
     procedure scrollContentViewBoundsChanged(notify: NSNotification); message 'scrollContentViewBoundsChanged:';
     procedure resetScrollRect; message 'resetScrollRect';
 
@@ -648,7 +646,7 @@
   Result:=documentView;
 end;
 
-procedure TCocoaScrollView.setDocumentView(aView: {$if FPC_FULLVERSION < 30200}NSView{$else}id{$endif});
+procedure TCocoaScrollView.setDocumentView(aView:NSView);
 begin
   inherited setDocumentView(aView);
   resetScrollRect;
