--- fpmake.pp	Wed Oct  6 08:46:33 2021
+++ fpmake.pp	Wed Oct  6 08:45:21 2021
@@ -149,6 +149,14 @@
     P.UnitPath.Add('compiler/packages/fcl-js/src');
     P.UnitPath.Add('compiler/packages/fcl-json/src');
     P.UnitPath.Add('compiler/packages/fcl-web/src/base');
+    P.UnitPath.Add('compiler/packages/fcl-net/src');
+    P.IncludePath.Add('compiler/packages/fcl-net/src/unix',AllUnixOSes);
+    P.IncludePath.Add('compiler/packages/fcl-net/src/win',AllWindowsOSes);
+    P.IncludePath.Add('compiler/packages/fcl-net/src/os2',[EMX]);
+    P.IncludePath.Add('compiler/packages/fcl-net/src/amiga',[morphos]);
+    P.IncludePath.Add('compiler/packages/fcl-net/src/$(OS)',AllOSes-AllWindowsOSes-AllUnixOSes-[EMX]);
+    P.UnitPath.Add('compiler/packages/fcl-base/src');
+    P.IncludePath.Add('compiler/packages/fcl-base/src/dummy');
 {$IFDEF VER3_0}    
     P.UnitPath.Add('compiler/packages/compat');
 {$ENDIF}    
