Index: compiler/systems/t_aros.pas
===================================================================
--- compiler/systems/t_aros.pas	(revision 44364)
+++ compiler/systems/t_aros.pas	(working copy)
@@ -92,7 +92,7 @@
     {       properly when calling the underlying GNU LD. (FIXME?)       }
     { This means paths with spaces in them are not supported for now on AROS.   }
     { So for example no Ram Disk: usage for anything which must be linked. (KB) }
-    ExeCmd[1]:='collect-aros $OPT $GCSECTIONS $ENTRY -d -n -o $EXE $RES';
+    ExeCmd[1]:='ld $OPT $GCSECTIONS $ENTRY -d -n -o $EXE $RES';
     ExeCmd[2]:='strip --strip-unneeded $EXE';
   end;
 end;