--- compiler/utils/fpc.pp
+++ compiler/utils/fpc.pp
@@ -122,8 +122,8 @@ program fpc;
      i : longint;
      errorvalue     : Longint;
   begin
-     setlength(ppccommandline,paramcount);
-     ppccommandlinelen:=0;
+     ppccommandlinelen:=2;
+     setlength(ppccommandline,paramcount+ppccommandlinelen);
      cpusuffix     :='';        // if not empty, signals attempt at cross
                                 // compiler.
      extrapath     :='';
@@ -183,7 +183,7 @@ program fpc;
      versionstr:='';                      { Default is just the name }
      if ParamCount = 0 then
        begin
-         SetLength (PPCCommandLine, 1);
+         SetLength (PPCCommandLine, PPCCommandLineLen+1);
          PPCCommandLine [PPCCommandLineLen] := '-?F' + ParamStr (0);
          Inc (PPCCommandLineLen);
        end
@@ -292,6 +292,7 @@ program fpc;
                 end;
             end;
        end;
+
      SetLength(ppccommandline,ppccommandlinelen);
 
      if versionstr<>'' then
@@ -308,6 +309,10 @@ program fpc;
             end;
         end;
 
+     ppccommandline[0]:='-n';
+     ppccommandline[1]:='@'+ExtractFilePath(ppcbin)+'fpc.cfg';
+     //ppccommandline[1]:='@'+ExtractFilePath(ParamStr(0))+'fpc.cfg';
+
      { call ppcXXX }
      try
        errorvalue:=ExecuteProcess(ppcbin,ppccommandline);
